doc: Add documentation and examples for mongodbatlas_metric_integration (#4561)
mongodb/terraform-provider-mongodbatlas: examples/mongodbatlas_metric_integration/README.md

# MongoDB Atlas Metric Integration Example

> This directory contains an example demonstrating how to configure an OpenTelemetry metric integration to export MongoDB Atlas metrics to an OTLP-compatible endpoint such as Datadog, New Relic, or Dynatrace.

Reviewer: The example is very datadog-focused. Should we make it clear here and move the example under a `/datadog` dir?
I see value in having a separate, more generic example as well.

Author: Good point on the rename to `/datadog`. However, regarding the example: `provider_type` only supports `CUSTOM`, `DYNATRACE`, and `NEW_RELIC`. Datadog isn't a native type, so the Datadog example already uses `CUSTOM`. 
A generic example would therefore also be CUSTOM and end up nearly identical to the Datadog one, just without the datadog provider resource. The only way to make a second example different would be to use a native type like NEW_RELIC or DYNATRACE

Reviewer: No issue with similar examples from my side, but up to you if you think its worth it in this case!

Author: See comment https://github.com/mongodb/terraform-provider-mongodbatlas/pull/4561#discussion_r3559548206

## Example

### [Metric Integration](.)

Configures a metric integration at the project level and demonstrates how to use the singular and plural data sources to read back the integration configuration.

**Resources created:**

- MongoDB Atlas Project
- MongoDB Atlas Metric Integration

## Prerequisites

- MongoDB Atlas Service Account with Organization Owner or Project Owner role.
- An OTLP-compatible endpoint URL and authentication credentials.

## Variables

| Variable | Description |
| --- | --- |

> | `atlas_org_id` | MongoDB Atlas Organization ID. |

Author: Did a bit of rework on the README in [fbf47bb](https://github.com/mongodb/terraform-provider-mongodbatlas/pull/4561/commits/fbf47bb961caa718a35481cef7725460cef8d77e)

> # MongoDB Atlas Metric Integration with OpenTelemetry Example

Reviewer 2: this follows the existing example pattern but i wonder if we need a standalone example directory for every resource, it's a tradeoff between showing functionality and adding noise (an example in the doc page might be sufficient in some cases)

Author: I'd lean toward keeping it.
 The setup is non-trivial, so a runnable project users can clone and apply de-risks adoption more than an inline snippet does, especially for a new resource where there's no prior config for people to copy. It also keeps us consistent with `log_integration`, the closest sibling in the OTel Sink family, which ships per-destination example dirs.
 Happy to hear feedback and if there are strong opinions from your side, @manupedrozo or @AgustinBettati 
 We tend to have an approach of creating new examples per resource for consistency, so if we change this principle I'd want to bring this up for team alignment.

Reviewer: I do prefer the targeted standalone examples per-resource even if there is overlap with other examples. It simplifies the lookup for customers - If I want an example for `metric_integration`, I know exactly where to find it.
We also use the example to feed into the doc generation so they play together nicely.

Author: Reworked the example to make it explicit it's for Datadog. The usage of different providers is fairly trivial as the config setup would be the same. Addressed the topic in [94bb9b2](https://github.com/mongodb/terraform-provider-mongodbatlas/pull/4561/commits/94bb9b21d005a51bd2b7f739e9114d3ce54ea4a6), let me know if it suits better!

# MongoDB Atlas Metric Integration with Datadog Example

This example demonstrates how to configure a metric integration to export MongoDB Atlas metrics to Datadog over OTLP. It also shows how to read the integration back with the singular and plural data sources.

> This example is focused on Datadog. To export to another provider, update `provider_type` (`CUSTOM`, `NEW_RELIC`, or `DYNATRACE`), the `endpoint`, and the authentication `headers` to match that provider.

Reviewer 3: ```suggestion
This example is specific to Datadog. To export to another provider, update `provider_type` (`CUSTOM`, `NEW_RELIC`, or `DYNATRACE`), the `endpoint`, and the authentication `headers` to match that provider.
```

Author: Addressed in [8c41a6e](https://github.com/mongodb/terraform-provider-mongodbatlas/pull/4561/commits/8c41a6e572fad977c32f368970e4a45ac10e530a)

## Prerequisites

- MongoDB Atlas Service Account with Organization Owner or Project Owner role.
- An OTLP-compatible endpoint URL and authentication credentials. This example uses Datadog and creates the Datadog API key with the `datadog` provider.

## Resources Created

This example creates the following resources:

### MongoDB Atlas

> - Project.

Reviewer 3: This should be more specific.

Author: Addressed in [8c41a6e](https://github.com/mongodb/terraform-provider-mongodbatlas/pull/4561/commits/8c41a6e572fad977c32f368970e4a45ac10e530a)

- Metric Integration configuration.

### Datadog

- API key used to authenticate metric ingestion.

## Usage

**1\. Ensure your MongoDB Atlas and Datadog credentials are set up.**

> This can be done using environment variables:

Reviewer 3: I suppose they can also be hardcoded, but this should be a simple instruction. "Set the following environment variables" or similar.

Author: Addressed in [8c41a6e](https://github.com/mongodb/terraform-provider-mongodbatlas/pull/4561/commits/8c41a6e572fad977c32f368970e4a45ac10e530a)

```bash
export MONGODB_ATLAS_CLIENT_ID="<ATLAS_CLIENT_ID>"
export MONGODB_ATLAS_CLIENT_SECRET="<ATLAS_CLIENT_SECRET>"
export DD_API_KEY="<DATADOG_API_KEY>"
export DD_APP_KEY="<DATADOG_APP_KEY>"
```

... or follow as in the `variables.tf` file and create a **terraform.tfvars** file with all the variable values:

```hcl
atlas_org_id        = "your-org-id"
atlas_client_id     = "your-service-account-client-id"
atlas_client_secret = "your-service-account-client-secret"
datadog_api_key     = "your-datadog-api-key"
datadog_app_key     = "your-datadog-app-key"
datadog_endpoint    = "https://otlp.datadoghq.com/v1/metrics"
```

**2\. Review the Terraform plan.**

> Execute the following command and ensure you agree with the plan.

Reviewer 3: ```suggestion
Execute the following command and confirm the plan's accuracy.
```

Author: Addressed in [8c41a6e](https://github.com/mongodb/terraform-provider-mongodbatlas/pull/4561/commits/8c41a6e572fad977c32f368970e4a45ac10e530a)

```bash
terraform plan
```

**3\. Execute the Terraform apply.**

Now execute the plan to provision the resources.

```bash
terraform apply
```

**4\. Destroy the resources.**

> When you have finished your testing, ensure you destroy the resources to avoid unnecessary Atlas charges.

Reviewer 3: ```suggestion
When you have finished your testing, ensure you destroy the resources to avoid accruing charges in Atlas.
```

Author: Addressed in [8c41a6e](https://github.com/mongodb/terraform-provider-mongodbatlas/pull/4561/commits/8c41a6e572fad977c32f368970e4a45ac10e530a)
